Source/WebCore/animation/TimelineRange.cpp

```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
     return value.valueID() == CSSValueNormal || (value.isPercentage() && !value.isCalculated() && percentIsDefault(value.resolveAsPercentageNoConversionDataRequired(), type));
 }
 
+Length SingleTimelineRange::defaultValue(Type type)
+{
+    return type == Type::Start ? Length(0, LengthType::Percent) : Length(100, LengthType::Percent);
+}
+
 bool SingleTimelineRange::isOffsetValue(const CSSPrimitiveValue& value)
 {
     return value.isLength() || value.isPercentage() || value.isCalculatedPercentageWithLength();
 }
 
+TimelineRangeValue SingleTimelineRange::serialize() const
+{
+    return TimelineRangeOffset { CSSPrimitiveValue::create(valueID(name))->stringValue(), offset.isPercentOrCalculated() ? CSSNumericFactory::percent(offset.percent()) : CSSNumericFactory::px(offset.value()) };
+}
+
+static const std::optional<CSSToLengthConversionData> cssToLengthConversionData(RefPtr<Element> element)
+{
+    CheckedPtr elementRenderer = element->renderer();
+    if (!elementRenderer)
+        return std::nullopt;
+    CheckedPtr elementParentRenderer = elementRenderer->parent();
+    Ref document = element->document();
+    CheckedPtr documentElement = document->documentElement();
+    if (!documentElement)
+        return std::nullopt;
+
+    return CSSToLengthConversionData { elementRenderer->style(), documentElement->renderer() ? &documentElement->renderer()->style() : nullptr, elementParentRenderer ? &elementParentRenderer->style() : nullptr, document->renderView() };
```

Comment:
Would be easier to read like this:
```suggestion
    return CSSToLengthConversionData {
        elementRenderer->style(),
        documentElement->renderer() ? &documentElement->renderer()->style() : nullptr,
        elementParentRenderer ? &elementParentRenderer->style() : nullptr, 
        document->renderView()
    };
```

Also, do container query units work? It would be useful to pass the element for container query resolution (or add a FIXME to fix it)

Reply:
Fixed.